### PR TITLE
RavenDB-20565 - allow the caller of PurgeAllCaches to dispose the old StringIndexCache

### PR DESCRIPTION
**Pull request by grisha-kotler**
https://issues.hibernatingrhinos.com/issue/RavenDB-20565/AccessViolationException-for-a-query-with-order-by

### CODE REVIEW

File: src/Lucene.Net/Search/FieldCacheImpl.cs [hunk 1]
```diff
@@ -73,11 +74,17 @@ private void  Init()
...
         {
             lock (this)
             {
+                // PurgeAllCaches is replacing the cache with a new one (without actually releasing any memory).
+                // When the GC will run, the finalizer of the Segments will be executed and release the unmanaged memory.
+                // We'll return the old StringIndexCache and let the caller decide if he wants to dispose it sooner
+
+                var copy = _stringIndexCache;
                 Init();
+                return copy;
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldCacheImpl.cs:87]: Do *not* return it, in this case.
Returning it make it look like someone will dispose it, which is likely not the desirable case(someone may be using this)

File: src/Lucene.Net/Search/FieldCacheImpl.cs [hunk 2]
```diff
@@ -829,6 +836,20 @@ protected internal override StringIndex CreateValue(IndexReader reader, Entry en
                 StringIndex value_Renamed = new StringIndex(retArray, retArrayOrdered, mterms);
                 return value_Renamed;
             }
+
+            public void Dispose()
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldCacheImpl.cs:840]: This won't be called from finalizer, no?

File: src/Lucene.Net/Search/FieldCacheImpl.cs [hunk 3]
```diff
@@ -829,6 +836,20 @@ protected internal override StringIndex CreateValue(IndexReader reader, Entry en
...
+            {
+                foreach (var keyValue in readerCache)
+                {
+                    foreach (var keyValuePair in keyValue.Value)
+                    {
+                        if (keyValuePair.Value is StringIndex si)
+                        {
+                            si.lookup.Dispose();
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldCacheImpl.cs:848]: If this is from finalizer, shouldn't we do that int the `StringIndex`, instead?

File: src/Lucene.Net/Search/FieldCacheImpl.cs [hunk 4]
```diff
@@ -73,11 +74,17 @@ private void  Init()
...
+        public virtual IDisposable PurgeAllCaches()
         {
             lock (this)
             {
+                // PurgeAllCaches is replacing the cache with a new one (without actually releasing any memory).
+                // When the GC will run, the finalizer of the Segments will be executed and release the unmanaged memory.
+                // We'll return the old StringIndexCache and let the caller decide if he wants to dispose it sooner
+
+                var copy = _stringIndexCache;
                 Init();
```
Comment (ayende) [inline src/Lucene.Net/Search/FieldCacheImpl.cs:86]: Pretty sure that we aren't locking on `this` on each read, so note that the `lock` isn't actually meaningful.
You may be getting partial reads here.


In this scenario, I don't think it matters, but something to be aware of

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Lucene.Net/Search/FieldCacheImpl.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Lucene.Net/Search/FieldCacheImpl.cs:86]: I want to lock at the RavenDB level, not here.
When I'm not running any queries.

Comment (ayende) [reply src/Lucene.Net/Search/FieldCacheImpl.cs:86]: who else is doing the locking?

Comment (grisha-kotler) [reply src/Lucene.Net/Search/FieldCacheImpl.cs:86]: This was the original code in Lucene. There is no need to do the locking if we are the only caller.

The idea is that the `Init` will be called when we aren't running any query.
The lock will be on the RavenDB side.

https://github.com/ravendb/ravendb/pull/16646/files#diff-3760b1669cd9af0aab370cfc92d98e3898fed9cea969a44cb19e7430aa953509R37

https://github.com/ravendb/ravendb/pull/16646/files#diff-3760b1669cd9af0aab370cfc92d98e3898fed9cea969a44cb19e7430aa953509R46

File: src/Lucene.Net/Search/FieldCacheImpl.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Lucene.Net/Search/FieldCacheImpl.cs:87]: That's the idea.
There is no way to tell if someone is currently using that cache.
From the RavenDB side, I can tell if we are running a query.
https://github.com/ravendb/ravendb/pull/16646/files#diff-
3760b1669cd9af0aab370cfc92d98e3898fed9cea969a44cb19e7430aa953509R37

File: src/Lucene.Net/Search/FieldCacheImpl.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Lucene.Net/Search/FieldCacheImpl.cs:840]: No, the finalizer is called for the `Segment`

File: src/Lucene.Net/Search/FieldCacheImpl.cs (hunk 4 above)
Comment (ayende) [reply src/Lucene.Net/Search/FieldCacheImpl.cs:86]: LGTM, then

**Review by maciejaszyk — APPROVED** [review]

**Pull request merged**